### PR TITLE
Add stall aware projectors (CORE-707)

### PR DESCRIPTION
**Pull request by rvesse**
Adds a new `StallAwareProjector` API into `projector-driver` module so that projectors that need to do something when projection stalls, i.e. no new events are available, can react accordingly.

As part of this refactored and cleaned up the `ProjectorDriver` test suite to cover more cases, add verification that certain log statements are triggered etc.

# Related Issues and PRs

- This is part of work towards addressing CORE-707, an edge case where a transaction batch was kept open indefinitely due to a slow low volume producer

### CODE REVIEW

File: projector-driver/pom.xml [hunk 1]
```diff
@@ -67,6 +67,11 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>com.github.valfirst</groupId>
+            <artifactId>slf4j-test</artifactId>
+        </dependency>
```
Comment (rvesse) [inline projector-driver/pom.xml:73]: - [ ] Should be `test` scope

File: projector-driver/src/test/java/io/smart/cache/projectors/driver/TestProjectorDriver.java [hunk 2]
```diff
@@ -68,8 +80,52 @@ public void cleanup() {
...
+    }
+
+    private static void waitExpectingError(Future<?> future, int timeout, TimeUnit timeUnit,
+                                           Class<?> expectedException) {
+        try {
+            future.get(timeout, timeUnit);
+            Assert.fail("Expected an error of type " + expectedException);
+        } catch (ExecutionException e) {
+            if (!expectedException.isAssignableFrom(e.getCause().getClass())) {
+                Assert.fail("Expected an error of type " + expectedException + " but got " + e.getClass());
```
Comment (rvesse) [inline projector-driver/src/test/java/io/smart/cache/projectors/driver/TestProjectorDriver.java:98]: - [ ] Message should reference `e.getCause().getClass()`

**Review by robwtelicent — APPROVED** [review]

**Comment by rvesse** [comment]
Fixed the failing build due to misformatted license headers

**Review by robwtelicent — APPROVED** [review]

**Review by TelicentPaul — APPROVED** [review]
Looks good

**Comment by rvesse** [comment]
I've fixed the flaky test by changing the test criteria slightly, and adding some retries to the offending test since it has some inherent randomness in it

**Review by robwtelicent — APPROVED** [review]
